File: src/sys/dotnet/fan/sys/StrBufOutStream.cs
//
// Copyright (c) 2007, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   12 Apr 07  Andy Frank  Creation
//

using System.Text;

namespace Fan.Sys
{
  /// <summary>
  /// StrBufOutStream.
  /// </summary>
  public class StrBufOutStream : SysOutStream
  {

  //////////////////////////////////////////////////////////////////////////
  // Constructor
  //////////////////////////////////////////////////////////////////////////

    public StrBufOutStream(StrBuf buf) : base(null)
    {
      m_sb = buf.sb;
      m_charsetEncoder = strBufEncoder();
    }

    public StrBufOutStream() : base(null)
    {
      m_sb = new StringBuilder();
      m_charsetEncoder = strBufEncoder();
    }

  //////////////////////////////////////////////////////////////////////////
  // OutStream
  //////////////////////////////////////////////////////////////////////////

    public string @string() { return m_sb.ToString(); }

  //////////////////////////////////////////////////////////////////////////
  // OutStream
  //////////////////////////////////////////////////////////////////////////

    public override OutStream w(int v)
    {
      throw UnsupportedErr.make("binary write on StrBuf output").val;
    }

    public override OutStream write(long x)
    {
      throw UnsupportedErr.make("binary write on StrBuf output").val;
    }

    public override OutStream writeBuf(Buf buf, long n)
    {
      throw UnsupportedErr.make("binary write on StrBuf output").val;
    }

    public override OutStream writeChar(char c)
    {
      m_sb.Append(c);
      return this;
    }

    public override OutStream writeChar(long c)
    {
      m_sb.Append((char)c);
      return this;
    }

    public override OutStream writeChars(string s, int off, int len)
    {
      m_sb.Append(s, off, len);
      return this;
    }

    public override OutStream flush()
    {
      return this;
    }

    public override bool close()
    {
      return true;
    }

  //////////////////////////////////////////////////////////////////////////
  // Charset
  //////////////////////////////////////////////////////////////////////////

    static Charset.Encoder strBufEncoder()
    {
      if (m_strBufEncoder == null) m_strBufEncoder = new StrBufEncoder();
      return m_strBufEncoder;
    }

    static Charset.Encoder m_strBufEncoder;

    class StrBufEncoder : Charset.Encoder
    {
      public override void encode(char ch, OutStream output)
      {
        ((StrBufOutStream)output).m_sb.Append(ch);
      }

      public override void encode(char ch, InStream input)
      {
        throw UnsupportedErr.make("binary write on StrBuf output").val;
      }
    }

  //////////////////////////////////////////////////////////////////////////
  // Fields
  //////////////////////////////////////////////////////////////////////////

    internal StringBuilder m_sb;

  }
}

File: src/sys/java/fan/sys/StrBufOutStream.java
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   1 Aug 06  Brian Frank  Creation
//
package fan.sys;

import java.io.*;

/**
 * StrBufOutStream
 */
public class StrBufOutStream
  extends OutStream
{

//////////////////////////////////////////////////////////////////////////
// Constructor
//////////////////////////////////////////////////////////////////////////

  public StrBufOutStream(StrBuf buf)
  {
    this.sb = buf.sb;
    this.charsetEncoder = strBufEncoder;
  }

  public StrBufOutStream()
  {
    this.sb = new StringBuilder();
    this.charsetEncoder = strBufEncoder;
  }

//////////////////////////////////////////////////////////////////////////
// OutStream
//////////////////////////////////////////////////////////////////////////

  public String string() { return sb.toString(); }

//////////////////////////////////////////////////////////////////////////
// OutStream
//////////////////////////////////////////////////////////////////////////

  public OutStream w(int v)
  {
    throw UnsupportedErr.make("Binary write on StrBuf.out").val;
  }

  public OutStream write(long x)
  {
    throw UnsupportedErr.make("Binary write on StrBuf.out").val;
  }

  public OutStream writeBuf(Buf buf, long n)
  {
    throw UnsupportedErr.make("Binary write on StrBuf.out").val;
  }

  public OutStream writeChar(char c)
  {
    sb.append(c);
    return this;
  }

  public OutStream writeChar(long c)
  {
    sb.append((char)c);
    return this;
  }

  public OutStream writeChars(String s, int off, int len)
  {
    sb.append(s, off, off+len);
    return this;
  }

  public OutStream flush()
  {
    return this;
  }

  public boolean close()
  {
    return true;
  }

  static final Charset.Encoder strBufEncoder = new Charset.Encoder()
  {
    public void encode(char ch, OutStream out)
    {
      ((StrBufOutStream)out).sb.append(ch);
    }

    public void encode(char ch, InStream out)
    {
      throw UnsupportedErr.make("Binary write on StrBuf.out").val;
    }
  };

//////////////////////////////////////////////////////////////////////////
// Fields
//////////////////////////////////////////////////////////////////////////

  StringBuilder sb;

}